python/opae.admin/opae/admin/tools/fpgareg.py

```diff
@@ -0,0 +1,367 @@
+# CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import re
+import os
+import argparse
+import struct
+import mmap
+import mmap
+import sys
+import binascii
+import glob
+import stat
+import time
+from ctypes import c_uint64, Structure, Union, c_uint32
+
+
+MAPSIZE = mmap.PAGESIZE
+MAPMASK = MAPSIZE-1
+
+FPGA_OFS_DEVID = ["0xbcce", "0xbccf", "0xaf00"]
+
+PATTERN = (r'.*(?P<segment>\w{4}):(?P<bus>\w{2}):'
+           r'(?P<dev>\w{2})\.(?P<func>\d).*')
+
+BDF_PATTERN = re.compile(PATTERN, re.IGNORECASE)
+
+HSSI_FEATURE_ID = 0x15
+PCIE_DFH_CSR_LEN = 30
+HSSI_DFH_CSR_LEN = 43
+ACCELERATOR_CSR_LEN = 30
+
+
+def verify_pcie_address(pcie_address):
+    """checks format of PCIe address string of the form bb:dd.f or
+    ssss:bb:dd.f, Returns true if  PCIe address of the form ssss:bb:dd.f,
+    or False if addr is not formatted correctly.
+    """
+    m = BDF_PATTERN.match(pcie_address)
+    if m is None:
+        print("Invalid pcie address foramt", pcie_address)
+        return False
+    return True
+
+
+def verify_fpga_device(pcie_address):
+    """checks PCIe address device id in FPGA Device id's list
+    Returns true if found in list, or
+    False if not found in device list.
+    """
+    paths = glob.glob(os.path.join("/sys/bus/pci/devices/",
+                                   pcie_address,
+                                   "device*"))
+    for path in paths:
+        with open(path, 'r') as fd:
+            devic_id = fd.read().strip()
+            if devic_id in FPGA_OFS_DEVID:
+                return True
+    return False
+
+
+class dfh_bits(Structure):
+    """
+    pcie device Feature Header bits
+    """
+    _fields_ = [
+                    ("id", c_uint64, 12),
+                    ("revision", c_uint64, 4),
+                    ("next_header_offset", c_uint64, 24),
+                    ("eol", c_uint64, 1),
+                    ("reserved", c_uint64, 19),
+                    ("type", c_uint64, 4)
+    ]
+
+
+class dfh(Union):
+    """
+    PCIE Device Feature Header
+    """
+    _fields_ = [("bits", dfh_bits),
+                ("value", c_uint64)]
+
+    def __init__(self, value):
+        self.value = value
+
+    @property
+    def id(self):
+        return self.bits.id
+
+    @property
+    def revision(self):
+        return self.bits.revision
+
+    @property
+    def next_header_offset(self):
+        return self.bits.next_header_offset
+
+    @property
+    def eol(self):
+        return self.bits.eol
+
+    @property
+    def type(self):
+        return self.bits.type
+
+
+class FPGAREG(object):
+    """
+    FPGAREG reads pcie,hssi,bms resoruces and dumps registers.
+    """
+    def __init__(self, args):
+        self._pcie_address = args
+        self.handle = None
+        self.mm = None
+
+    def open(self, address, pcie_address=None):
```

Comment:
Have you considered using the \_\_enter__ and \_\_exit__ pattern so that you can use with?  It helps eliminate a missed .close() call.